File: solutions/java/0712-minimum-ascii-delete-sum-for-two-strings.java
class Solution {
    public int minimumDeleteSum(String s1, String s2) {
        int[][] dp = new int[s1.length() + 1][s2.length() + 1];

        for (int i = 1; i <= s1.length(); i++)
            dp[i][0] = dp[i - 1][0] + s1.charAt(i - 1);
        
        for (int j = 1; j <= s2.length(); j++)
            dp[0][j] = dp[0][j - 1] + s2.charAt(j - 1);
        
        for (int i = 1; i <= s1.length(); i++) {
            for (int j = 1; j <= s2.length(); j++) {
                if (s1.charAt(i - 1) == s2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = Math.min(dp[i - 1][j] + s1.charAt(i - 1), dp[i][j - 1] + s2.charAt(j - 1));
                }
            }
        }
        
        return dp[s1.length()][s2.length()];
    }
}

File: solutions/rust/0712-minimum-ascii-delete-sum-for-two-strings.rs
impl Solution {
    pub fn minimum_delete_sum(s1: String, s2: String) -> i32 {
        let mut prev_row = vec![0; s2.len() + 1];
        let s1 = s1.chars().collect::<Vec<_>>();
        let s2 = s2.chars().collect::<Vec<_>>();

        for j in 1..=s2.len() {
            prev_row[j] = prev_row[j - 1] + s2[j - 1] as i32;
        }

        for i in 1..=s1.len() {
            let mut curr_row = vec![prev_row[0] + s1[i - 1] as i32];
            for j in 1..=s2.len() {
                if s1[i - 1] == s2[j - 1] {
                    curr_row.push(prev_row[j - 1]);
                } else {
                    curr_row.push(std::cmp::min(prev_row[j] + s1[i - 1] as i32, curr_row[j - 1] + s2[j - 1] as i32));
                }
            }
            prev_row = curr_row;
        }

        *prev_row.last().unwrap()
    }
}
